### PR TITLE
Enable native staging apps to use Staging API

### PR DESCRIPTION
**Pull request by mountiny**
### Details

This PR will enable us to use staging API in native apps as well. This is a follow up to this PR https://github.com/Expensify/App/pull/12880 which had to be reverted as after one refactor update the dev environment was not read correctly locally.

The shouldUseStaging variable will be false by default in native apps. In production App, user does not have access to the toggle hence they cannot switch production app to use Staging API. In staging, this means that before the Onyx data is available we will defer to production API, which is safe bet and how staging app works now. Once the settings is available we will decide based on the toggle state.

For desktop and web, we will keep the same behaviour which we achieve by using the `getPlatfrom` call to check for Web and desktop.

In case of Android and iOS, the Config file is not accessible in staging build (we use production env during building), therefore we need to rely only on the staging toggle in this case and that is why we add this platform specific check.

### Fixed Issues

$ https://github.com/Expensify/App/issues/11561
$ https://github.com/Expensify/App/issues/11068

### Tests

- [x] Verify that no errors appear in the JS console

Internal tests

1. Copy the contents from the `.env.production` to your `.env` file
2. Make changes to [this line](https://github.com/Expensify/App/blob/eb0110d8d2bc2d1ed00828a1c9efdcbb6a8929d7/src/pages/settings/PreferencesPage.js#L103) so the test preferences are shown in 
3. Run the iOS and Android app
4. Make sure the toggle for staging api is turned on
5. Performs some actions in the app
6. Go to logs and search for the most recent actions by this account
7. Make sure the servers these request hit are staging servers
8. Go back to the settings and switch to production API
9. Repeat and confirm in the logs the request are hitting the production servers

1. Run the app in desktop with staging env
2. Confirm the correct servers are hit based on the toggle value

1. Run the app in desktop with production env
2. Confirm the production servers are used

#### Test that local development still works

1. Run with dev env and make sure the API hits your local VM

### Offline tests

No offline tests here.

### QA Steps

- [ ] Verify that no errors appear in the JS console

1. Open the iOS and Android staging app
1. Make sure the toggle for staging api is turned on
1. Performs some actions in the app
1. Go to logs and search for the most recent actions by this account
1. Make sure the servers these request hit are staging servers
1. Go back to the settings and switch to production API
1. Repeat and confirm in the logs the request are hitting the production servers

1. Run the staging web
2. By checking the logs, confirm the correct servers are being hit based on what stage the toggle is at

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).

- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for expected offline behavior are in the `Offline steps` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots/Videos

1. Using dev env confirming that is not broken:
- in web:
Screenshot 2022-12-05 at 13 42 23

- in native also seems to point ot dev just fine:
image

2. Staging env on desktop
- 
image

- staging but toggling to use production api works well too
image

5. Staging env on native (that is how I can test locally the staging switch as the local build with production API is considered to be production app so no toggle shows up) - for this I had to force staging environment also [here](https://github.com/Expensify/App/blob/main/src/libs/Environment/getEnvironment/index.native.js#L28) as the badge and the settings rely on request to play store to check the App version which obviously does not work locally.
image

the production api has been used if the toggle has changed back to prodcution

6. Using Production API

- desktop 
image
 - native using production API correctly sends the requests to production
image

<details>
<summary>Web</summary>

image

with staging variables 
image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

image

image

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by mountiny** [comment]
This is WIP because there will be still an issue now with Production native app pointing to staging.

After thinking about this for a while I think we dont have other option than to default the staging api toggle to false and hence start the staging apps with production API until the user changes this settings to staging API (in native at least). 

I think it is not ideal but I think this is fine as it does not stop us from testing this.

**Comment by marcochavezf** [comment]
> In case of Android and iOS, the Config file is not accessible in staging build (we use production env during building)

@mountiny, I think this is why `IS_IN_STAGING` is not setting the correct value, but we're already able to know if the env is staging on native apps (i.e. Android app from the internal beta program). For example, we display the [TestToolMenu](https://github.com/Expensify/App/blob/a34851568e3f004f3bd8b6aff66102c91f97593f/src/pages/settings/PreferencesPage.js#L103) and the [staging logo](https://github.com/Expensify/App/blob/98b72b12fd563eadd01d806d30f2b93a380a2c13/src/components/ExpensifyCashLogo.js) on staging native apps. So what if we use the same logic of the TestToolMenu or the staging log to use the Staging API in HttpUtils?

**Comment by mountiny** [comment]
> because environment === CONST.ENVIRONMENT.STAGING is true (expected value) in other parts of the staging native apps

In other parts of the staging native apps we always use [getEnvironment](https://github.com/Expensify/App/blob/c10d5fb2d5c196986cb614cd2161ee83577ebcba/src/libs/Environment/getEnvironment/index.native.js#L1-L35) method which makes the API request

**Comment by marcochavezf** [comment]
> In other parts of the staging native apps we always use [getEnvironment](https://github.com/Expensify/App/blob/c10d5fb2d5c196986cb614cd2161ee83577ebcba/src/libs/Environment/getEnvironment/index.native.js#L1-L35) method which makes the API request

Oh I see, but is there a problem if we block the other requests until the [getEnvironment](https://github.com/Expensify/App/blob/c10d5fb2d5c196986cb614cd2161ee83577ebcba/src/libs/Environment/getEnvironment/index.native.js#L1-L35) is completed? 

>Therefore the best solution which will work predictably and consistently is to default to production instead of staging and then just switch based on the toggle once onyx state if loaded (in case on native apps).

Sorry if I'm missing something, but how will we know that we have to turn on the staging switch automatically on staging apps? Isn't more reliable to wait until `getEnvironment` is completed before we make any other request?

**Comment by mountiny** [comment]
> is there a problem if we block the other requests until the [getEnvironment](https://github.com/Expensify/App/blob/c10d5fb2d5c196986cb614cd2161ee83577ebcba/src/libs/Environment/getEnvironment/index.native.js#L1-L35) is completed?

Normally we dont call this as app start, which we would have to do here and also even in production since we dont have a way to know to not make this request if in production. If the request is made somewhere else in the App it is not such huge problem.

> Sorry if I'm missing something, but how will we know that we have to turn on the staging switch automatically on staging apps? Isn't more reliable to wait until getEnvironment is completed before we make any other request?

the getEnvironment is used to decide if we should show  the toggle, then if we are in staging, we show the toggle but its actual value is help only in Onyx and that is what will decide where we will point. We cannot however safely default to staging, we will need to default to production and then let the user who is on staging switch to staging. That will be saved until the onyx data is reset by signing out (I dont think we save this in database)

**Comment by mountiny** [comment]
@marcochavezf @parasharrajat Not sure how Rajat you can test this in native apps locally but this is ready for review/ testing.

I have included all my testing scenarios in the PR body, it took a while to test everything.

There is no ideal solution to this so I feel like this is the least evil route.

File: src/components/TestToolMenu.js [hunk 1]
```diff
@@ -42,7 +45,7 @@ const TestToolMenu = props => (
         This enables QA and internal testers to take advantage of sandbox environments for 3rd party services like Plaid and Onfido. */}
         <TestToolRow title="Use Staging Server">
             <Switch
-                isOn={lodashGet(props, 'user.shouldUseStagingServer', true)}
+                isOn={lodashGet(props, 'user.shouldUseStagingServer', _.contains([CONST.PLATFORM.WEB, CONST.PLATFORM.DESKTOP], getPlatform()))}
```
Comment (mountiny) [inline src/components/TestToolMenu.js:48]: Only default to true if we are in web or desktop which can read the correct config

**Comment by marcochavezf** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by marcochavezf — APPROVED** [review]
Thanks @mountiny! Tests went well

**Comment by mountiny** [comment]
Thank you @marcochavezf

**Comment by mountiny** [comment]
QAed this on web and iOS mobile and it works well 🎉

**Pull request merged**